src/com/cdd/bao/template/Schema.java

```diff
@@ -677,6 +687,17 @@ public static boolean compatiblePropGroupNest(String propURI1, String[] groupNes
 		return propURI1.equals(propURI2) && compatibleGroupNest(groupNest1, groupNest2);
 	}
 	
+	// convenience methods for combining parts of assignments/annotations to make string identifiers
+	private final static String SEP = "::";
+	public static String keyPropGroup(String propURI, String[] groupNest)
+	{
+		return propURI + SEP + (groupNest == null ? "" : String.join(SEP, groupNest));
```

Comment:
If we add a SEP at the end, it's obvious that we have a key that identifies a propURI and a groupNest. If there is no SEP at the end, the key contains a value.

- prop::group::
- prop::group::value

Reply:
I think it's better not to give the keys any intrinsic properties... I've used keys in a lot of places and none of them needed to resolve which kind. But if anyone runs into a use case, we could upgrade it without breaking existing functionality, and add an extra method for querying what type it is.